File: benches/bench_xbrl.rs
use criterion::{Criterion, criterion_group, criterion_main};
use quick_xml::Reader;
use std::{path::PathBuf, str::FromStr, time::Duration};
use xbrl_rs::{InstanceDocument, TaxonomySet};

const TAXONOMY_ENTRY_POINT: &str = "test_data/taxonomies";

fn parse_instance(c: &mut Criterion) {
    let mut group = c.benchmark_group("bench_xbrl");
    group.sample_size(10);
    group.measurement_time(Duration::from_secs(10));
    group.warm_up_time(Duration::from_secs(5));
    group.bench_function("parse_instance", |b| {
        b.iter(|| {
            let reader = Reader::from_file("test_data/instances/balance_sheet_v64.xml").unwrap();
            let _instance = InstanceDocument::from_xml(reader).unwrap();
        });
    });
}

fn validate_instance(c: &mut Criterion) {
    let entry_point = PathBuf::from_str(TAXONOMY_ENTRY_POINT).unwrap();
    let taxonomy = TaxonomySet::discover(
                vec![
                    "http://www.xbrl.de/taxonomies/de-gcd-2020-04-01/de-gcd-2020-04-01-shell.xsd".to_owned(),
                    "http://www.xbrl.de/taxonomies/de-gaap-ci-2020-04-01/de-gaap-ci-2020-04-01-shell-fiscal.xsd"
                        .to_owned(),
                ],
                entry_point,
            )
            .unwrap();
    let reader = Reader::from_file("test_data/instances/balance_sheet_v64.xml").unwrap();
    let instance = InstanceDocument::from_xml(reader).unwrap();

    let mut group = c.benchmark_group("bench_xbrl");
    group.sample_size(10);
    group.measurement_time(Duration::from_secs(10));
    group.warm_up_time(Duration::from_secs(5));
    group.bench_function("validate_instance", |b| {
        b.iter(|| instance.validate(&taxonomy));
    });
}

fn view_instance(c: &mut Criterion) {
    let entry_point = PathBuf::from_str(TAXONOMY_ENTRY_POINT).unwrap();
    let taxonomy = TaxonomySet::discover(
                vec![
                    "http://www.xbrl.de/taxonomies/de-gcd-2020-04-01/de-gcd-2020-04-01-shell.xsd".to_owned(),
                    "http://www.xbrl.de/taxonomies/de-gaap-ci-2020-04-01/de-gaap-ci-2020-04-01-shell-fiscal.xsd"
                        .to_owned(),
                ],
                entry_point,
            )
            .unwrap();
    let reader = Reader::from_file("test_data/instances/balance_sheet_v64.xml").unwrap();
    let instance = InstanceDocument::from_xml(reader).unwrap();

    let mut group = c.benchmark_group("bench_xbrl");
    group.sample_size(10);
    group.measurement_time(Duration::from_secs(10));
    group.warm_up_time(Duration::from_secs(5));
    group.bench_function("view_instance", |b| {
        b.iter(|| instance.validate(&taxonomy));
    });
}

fn bench_single_dts_2020(c: &mut Criterion) {
    // Single entry point benchmark.
    let schema_refs = vec![
        "http://www.xbrl.de/taxonomies/de-bra-2020-04-01/de-bra-2020-04-01-shell-fiscal.xsd"
            .to_owned(),
    ];

    let mut group = c.benchmark_group("bench_xbrl");
    group.sample_size(10);
    group.measurement_time(Duration::from_secs(60));
    group.warm_up_time(Duration::from_secs(5));
    group.bench_function("single_dts_2020", |b| {
        b.iter(|| {
            let entry_point = PathBuf::from_str(TAXONOMY_ENTRY_POINT).unwrap();
            TaxonomySet::discover(schema_refs.clone(), entry_point).unwrap()
        });
    });
    group.finish();
}

fn bench_full_dts_2020(c: &mut Criterion) {
    // All 6 entry points as used in a real HGB instance document.
    let schema_refs =     vec![
        "http://www.xbrl.de/taxonomies/de-gcd-2020-04-01/de-gcd-2020-04-01-shell.xsd".to_owned(),
        "http://www.xbrl.de/taxonomies/de-gaap-ci-2020-04-01/de-gaap-ci-2020-04-01-shell-fiscal.xsd"
            .to_owned(),
        "http://www.xbrl.de/taxonomies/de-bra-2020-04-01/de-bra-2020-04-01-shell-fiscal.xsd"
            .to_owned(),
        "http://www.xbrl.de/taxonomies/de-fi-2020-04-01/de-fi-2020-04-01-shell-staffelform-fiscal.xsd"
            .to_owned(),
        "http://www.xbrl.de/taxonomies/de-ins-2020-04-01/de-ins-2020-04-01-shell-fiscal.xsd"
            .to_owned(),
        "http://www.xbrl.de/taxonomies/de-pi-2020-04-01/de-pi-2020-04-01-shell-staffelform-fiscal.xsd"
            .to_owned(),
    ];

    let mut group = c.benchmark_group("bench_xbrl");
    group.sample_size(10);
    group.measurement_time(Duration::from_secs(60));
    group.warm_up_time(Duration::from_secs(5));
    group.bench_function("full_dts_2020", |b| {
        b.iter(|| {
            let entry_point = PathBuf::from_str(TAXONOMY_ENTRY_POINT).unwrap();
            TaxonomySet::discover(schema_refs.clone(), entry_point).unwrap()
        });
    });
    group.finish();
}

criterion_group!(
    benches,
    parse_instance,
    validate_instance,
    view_instance,
    bench_single_dts_2020,
    bench_full_dts_2020,
);
criterion_main!(benches);

File: benches/bench_xbrl.py
#!/usr/bin/env python3
# /// script
# requires-python = ">=3.11"
# dependencies = [
#   "arelle-release~=2.38.13",
# ]
# ///
"""
Benchmark for XBRL taxonomy discovery using Arelle.

Measures the time to load the full DTS for the 2020-04-01 German taxonomy
suites, for comparison with the Rust xbrl-rs benchmarks.

Usage:
    uv run benches/taxonomy_parsing.py
"""

import os
import statistics
import sys
import time

from arelle.ModelFormulaObject import FormulaOptions
from arelle import Cntlr
from arelle.ValidateXbrl import ValidateXbrl

REPO_ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
TAXONOMY_DIR = os.path.join(REPO_ROOT, "test_data", "taxonomies")
INSTANCE_PATH = os.path.join(
    REPO_ROOT, "test_data", "instances", "balance_sheet_v64.xml"
)

# All 6 entry points as used in a real HGB instance document.
SCHEMA_REFS_2020 = [
    os.path.join(TAXONOMY_DIR, "de-gcd-2020-04-01",
                 "de-gcd-2020-04-01-shell.xsd"),
    os.path.join(TAXONOMY_DIR, "de-gaap-ci-2020-04-01",
                 "de-gaap-ci-2020-04-01-shell-fiscal.xsd"),
    os.path.join(TAXONOMY_DIR, "de-bra-2020-04-01",
                 "de-bra-2020-04-01-shell-fiscal.xsd"),
    os.path.join(
        TAXONOMY_DIR,
        "de-fi-2020-04-01",
        "de-fi-2020-04-01-shell-staffelform-fiscal.xsd",
    ),
    os.path.join(TAXONOMY_DIR, "de-ins-2020-04-01",
                 "de-ins-2020-04-01-shell-fiscal.xsd"),
    os.path.join(
        TAXONOMY_DIR,
        "de-pi-2020-04-01",
        "de-pi-2020-04-01-shell-staffelform-fiscal.xsd",
    ),
]

# Single entry point benchmark.
SCHEMA_REFS_SINGLE_2020 = [
    os.path.join(TAXONOMY_DIR, "de-bra-2020-04-01",
                 "de-bra-2020-04-01-shell-fiscal.xsd"),
]

ITERATIONS = 5


def create_controller():

    controller = Cntlr.Cntlr(logFileName="logToStdErr")
    controller.startLogging(logFileName=None)
    return controller


def prepare_bench():
    """Set up the instance model once, return a closure that only measures validation."""
    controller = create_controller()
    opts = FormulaOptions()
    # Disable formula validation as it is not supported by xbrl-rs yet.
    opts.formulaAction = "none"
    controller.modelManager.formulaOptions = opts
    model = load_instance(controller)
    return lambda: validate_instance(controller, model)


def load_taxonomy_set(cntlr, schema_paths: list[str]) -> float:
    """Load all schemas sequentially, returning total elapsed seconds.

    Arelle builds a fully independent model per load() call with no cross-call
    deduplication, so each schema is timed and closed individually. This differs
    from TaxonomySet::discover(), which merges all entry points into one DTS and
    parses each file exactly once — making the Rust benchmark inherently faster for
    entry point sets that share schemas (de-bra and de-ins both import de-gaap-ci).
    """
    total = 0.0

    for path in schema_paths:
        url = "file://" + os.path.abspath(path).replace("\\", "/")
        t0 = time.perf_counter()
        model_xbrl = cntlr.modelManager.load(url)
        total += time.perf_counter() - t0
        if model_xbrl is not None:
            cntlr.modelManager.close(model_xbrl)

    return total


def load_instance(cntlr) -> "ModelXbrl":
    """Load the instance document, returning the model."""
    url = "file://" + os.path.abspath(INSTANCE_PATH).replace("\\", "/")
    return cntlr.modelManager.load(url)


def validate_instance(cntlr, model_xbrl) -> float:
    """Validate a loaded instance, returning elapsed seconds."""
    t0 = time.perf_counter()
    validator = ValidateXbrl(model_xbrl)
    validator.validate(model_xbrl)
    elapsed = time.perf_counter() - t0
    validator.close()
    return elapsed


def run_benchmark(label: str, bench_fn, iterations: int) -> None:
    print(f"\n{label}")
    print(f"  {'sample':>8}   {'time':>12}")
    samples: list[float] = []

    # Warm-up run (not recorded) to populate the OS file cache — mirrors
    # Criterion's warm_up_time before actual measurement begins.
    bench_fn()

    for i in range(iterations):
        elapsed = bench_fn()
        samples.append(elapsed)
        print(f"  {i + 1:>8}   {elapsed * 1000:>10.2f} ms")

    mean = statistics.mean(samples)
    stdev = statistics.stdev(samples) if len(samples) > 1 else 0.0
    print(f"  {'mean':>8}   {mean * 1000:>10.2f} ms  ± {stdev * 1000:.2f} ms")
    print(f"  {'min':>8}   {min(samples) * 1000:>10.2f} ms")
    print(f"  {'max':>8}   {max(samples) * 1000:>10.2f} ms")


def main() -> int:
    try:
        from arelle import Cntlr  # noqa: F401
    except ImportError:
        print(
            "ERROR: arelle is not installed. Run: uv run benches/taxonomy_parsing.py",
            file=sys.stderr,
        )
        return 1

    print("Arelle XBRL benchmark")
    print(f"Taxonomy directory : {TAXONOMY_DIR}")
    print(f"Instance document  : {INSTANCE_PATH}")
    print(f"Iterations         : {ITERATIONS}")

    run_benchmark(
        "taxonomy_discovery/full_dts_2020",
        lambda: load_taxonomy_set(create_controller(), SCHEMA_REFS_2020),
        ITERATIONS,
    )
    run_benchmark(
        "taxonomy_discovery/single_dts_2020",
        lambda: load_taxonomy_set(
            create_controller(), SCHEMA_REFS_SINGLE_2020),
        ITERATIONS,
    )

    run_benchmark("validate_instance", prepare_bench(), ITERATIONS)

    print()
    return 0


if __name__ == "__main__":
    sys.exit(main())
